xc/device_common.ml

```diff
@@ -338,22 +339,101 @@ let is_upstream_qemu domid =
     with_xs (fun xs -> xs.Xs.read (sprintf "/libxl/%d/dm-version" domid)) = "qemu_xen"
   with _ -> false
 
-let qmp_write_and_read domid ?(read_result=true) cmd  =
-  try
-    let open Qmp in
-    debug "QMP.Command for domid %d: %s" domid (string_of_message cmd);
-    let c = Qmp_protocol.connect (qmp_libxl_path domid) in
-    finally
-      (fun () ->
-         Qmp_protocol.negotiate c;
-         Qmp_protocol.write c cmd;
-         if read_result then
-           Some (Qmp_protocol.read c)
-         else None
-      )
-      (fun () -> Qmp_protocol.close c)
-  with e ->
-    error "Caught exception attempting to write qmp message: %s" (Printexc.to_string e);
-    None
+let qmp domid cmd  =
+  let open Qmp in
+  let cmd' = string_of_message cmd in
+  debug "QMP.Command for domid %d: %s" domid cmd';
+  let c = Qmp_protocol.connect (qmp_libxl_path domid) in
+  finally
+    (fun () ->
+       Qmp_protocol.negotiate c;
+       Qmp_protocol.write c cmd;
+       Qmp_protocol.read c
+    )
+    (fun () -> Qmp_protocol.close c)
+
+(** [make_id_generator] returns a function that creates unique IDs using
+ * an internal counter *)
+let make_id_generator name =
+  let count = ref 0 in
+  fun domid -> incr count; Printf.sprintf "%s-%06d-%d" name !count domid
+
+let make_qmp_id = make_id_generator "qmp"
+
+(** [qmp_send_cmd_internal connection omid cmd] sends [cmd] to QEMU, waits
+ * for the *matching* response, and returns the result. It uses an
+ * already open and negotiated [connection]. Commands are tagged with an
+ * identifier and the function waits for the matching response, skipping
+ * all other responses it receives while waiting.
+ *)
```

Comment:
Do we need a timeout for waiting?

Reply:
Maybe an optional timeout, some commands may take an indefinite time to finish if it involves a lot of I/O. Is it possible to cancel QMP commands after the timeout expires?